### PR TITLE
Handling unix socket connection based on OS

### PR DESCRIPTION
**Pull request by shrey-gang**
Pull Request is partenered with [asticode/go-astilectron#191](https://github.com/asticode/go-astilectron/pull/191)

This will fix issue: [asticode/go-astilectron#190](https://github.com/asticode/go-astilectron/issues/190)

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -85,6 +85,10 @@ app.on('ready',() => {
         switch (json.name) {
             // App
             case consts.eventNames.appCmdQuit:
+            rl.close()
+            setTimeout(function(){
```
Comment (asticode) [inline main.js]: Why do you add a timeout here? And won't it be cancelled by `app.quit()` just  below?

File: src/client.js [hunk 2]
```diff
@@ -22,6 +31,22 @@ class Client {
...
         this.socket.write(JSON.stringify(data) + "\n")
     }
+
+    // for proper socket closing
+    close() {
+        this.socket.end()
+    }
+
+    // getconnection establishes connection based on underlying OS
+    getconnection() {
```
Comment (asticode) [inline src/client.js]: Could you rename this function to `connect`?

File: src/client.js [hunk 3]
```diff
@@ -22,6 +31,22 @@ class Client {
...
+        if ( os.platform() != "win32" ) {
```
Comment (asticode) [inline src/client.js]: This should check `process.argv[2]` instead

File: src/client.js [hunk 4]
```diff
@@ -22,6 +31,22 @@ class Client {
...
+            this.socket = net.createConnection(process.argv[1]);
```
Comment (asticode) [inline src/client.js]: Why have you transform it to `process.argv[1]`? In the original file it's `process.argv[2]`. And now that we've added another flag, it should be `process.argv[3]`

File: src/client.js [hunk 5]
```diff
@@ -22,6 +31,22 @@ class Client {
...
+        } else {
+            let u = url.parse("tcp://" + process.argv[1], false, false)
```
Comment (asticode) [inline src/client.js]: Same here for `process.argv[1]`

**Review by asticode — CHANGES_REQUESTED** [review]
Something is weird with the flag index you've used.

By the way, since we've added an additionnal flag, all flags with their index >=2 should be (index +1).

Other changes are needed.

File: main.js (hunk 1 above)
Comment (shrey-gang) [reply main.js]: I was testing something around this, might have missed removing this code.

Comment (shrey-gang) [reply main.js]: removed code

File: src/client.js (hunk 2 above)
Comment (shrey-gang) [reply src/client.js]: renamed to connect()

File: src/client.js (hunk 3 above)
Comment (shrey-gang) [reply src/client.js]: rectified argument number

File: src/client.js (hunk 4 above)
Comment (shrey-gang) [reply src/client.js]: rectified argument number

File: src/client.js (hunk 5 above)
Comment (shrey-gang) [reply src/client.js]: rectified argument number

**Comment by asticode** [comment]
I had to merge another PR that moved some stuff around and now your PR conflicts with the `master` branch. Can you make sure there are no conflicts?

**Comment by shrey-gang** [comment]
I have resolved all the conflicts.

File: src/client.js [hunk 6]
```diff
@@ -5,23 +5,51 @@ const url = require("url");
...
-    if (typeof payload !== "undefined") Object.assign(data, payload);
-    this.socket.write(JSON.stringify(data) + "\n");
-  }
+    // init initializes the Client
+    init(addr) {
+
+        this.connect(addr)
+
+        this.socket.on('error', function(err){
+            // Writing to a file in case of error related to socket
```
Comment (asticode) [inline src/client.js]: In this case I'd rather throw an exception (that will be printed on the user's screen) and kill the app instead of writing to a file few people are aware of.

I don't know how Electron does it and I'm not really familiar with NodeJS, but when there's a fatal error in Electron, it prints the error in a nice window and exit the process. Could you do the same thing instead?

If this is not possible, you should only have to print the error in `stderr` which will then be picked up by the `go` side that will print it in its logs.

File: src/client.js [hunk 7]
```diff
@@ -5,23 +5,51 @@ const url = require("url");
...
+
+    // write writes an event to the server
+    write(targetID, eventName, payload) {
+        let data = {name: eventName, targetID: targetID}
+        if (typeof payload !== "undefined") Object.assign(data, payload)
+        this.socket.write(JSON.stringify(data) + "\n")
+    }
+
+    /*
+     * for proper socket closing, unix socket remains open even after
```
Comment (asticode) [inline src/client.js:40]: nice, I didn't know that!

File: src/client.js [hunk 8]
```diff
@@ -5,23 +5,51 @@ const url = require("url");
...
+     * quiting the application, this function ends the socket connection
+     */
+    close() {
+        this.socket.end()
+    }
+
+    // establishes connection based on underlying OS
+    connect(addr) {
+        if ( os.platform() != "win32" ) {
```
Comment (asticode) [inline src/client.js]: `addr` here should be the full `addr` hence `tcp://127.0.0.1:4003` or `unix://path/to/socket`.

That way, instead of creating the connection based on OS, it should be created based on the `addr` scheme (`tcp://` or `unix://`)

The `go` side will have to be updated as well.

**Review by asticode — CHANGES_REQUESTED** [review]

File: src/client.js (hunk 8 above)
Comment (shrey-gang) [reply src/client.js]: Going with this approach, we first have to check for "tcp://" or "unix://" and then split the string to get the "host: port" or "path" and then pass it to either socket.connect() or net.createConnection().

If we use net.createConnection(), then the possible arguments to this function are:
1. {port: 1234}
2. {path: '/tmp/xyz.sock'}

So my confusion is why don't we go with the first approach, as there are fewer checks to be done and the code/logic seems simpler.

@asticode it would be helpful if you can explain it with some details?

Comment (asticode) [reply src/client.js]: My problem with the first approach is that both GO and JS have the responsibility of determining the connection scheme based on the OS. As far as I'm concerned, that shouldn't be the case since only one side should have this logic. Otherwise, when we update such a logic, we'll have to update both GO and JS.

Hence I'd rather have GO driving the value (and scheme) of `addr` based on the OS and JS creating the proper connection based on the scheme of the received `addr`. That way reponsibilities are properly divided.

In a nutshell:

```
GO decides the value of addr based on the OS
|
GO sends the value of addr to JS through a flag
|
JS creates the proper connection based on the addr (and scheme) received
```

That way, if we decide to update the `addr` logic, we only have to update GO.

Does that make more sense?

Comment (shrey-gang) [reply src/client.js]: Yes, understood your point of view, I have made those changes as per your suggestions.
Testing those changes, hopefully I will push the changes soon.

File: main.js [hunk 9]
```diff
@@ -4,6 +4,7 @@ const app = require("electron");
 const start = require("./index").start;
 const lastWindow = require("./index").lastWindow;
 
+
```
Comment (asticode) [inline main.js]: Can you remove this empty line?

File: src/client.js [hunk 10]
```diff
@@ -1,27 +1,56 @@
...
+    /*
+     * for proper socket closing, unix socket remains open even after
+     * quiting the application, this function ends the socket connection
+     */
+    close() {
+        this.socket.end();
+    }
+
+    connect(addr) {
+        if (net.isIP(addr) > 0) {
```
Comment (asticode) [inline src/client.js]: I don't really like the idea of using `net.isIP` here.

Can you check whether simplifying to this works :

```js
this.socket = net.createConnection(addr);
```

If not, I'd rather see something along the lines of :

```js
let u = url.parse(addr, false, false);
if (u.protocol === "tcp:") {
  this.socket = new net.Socket();
  this.socket.connect(u.port, u.hostname, function() {});
} else {
  this.socket = net.createConnection(addr);
}
```

File: src/client.js [hunk 11]
```diff
@@ -1,27 +1,56 @@
...
-    if (typeof payload !== "undefined") Object.assign(data, payload);
-    this.socket.write(JSON.stringify(data) + "\n");
-  }
+    // init initializes the Client
+    init(addr) {
+        this.connect(addr)
+
+        this.socket.on('error', function(err){
+            // Raising an exception in case of any error in socket
+            const messageBoxOptions = {
```
Comment (asticode) [inline src/client.js:16]: Why do you need to show the error in a dialog ? Writing it in `stdout` or `stderr` should be sufficient since it should be picked up by the `go` side.

**Review by asticode — CHANGES_REQUESTED** [review]

File: src/client.js (hunk 11 above)
Comment (shrey-gang) [reply src/client.js:16]: I thought showing the message in a proper window is what you suggested in your earlier comment, so I made the changes accordingly plus showing a proper reason as to why the application closed/crashed abruptly will be a better option.

File: src/client.js (hunk 10 above)
Comment (shrey-gang) [reply src/client.js]: I tried with [below code], but there was some issue in windows with this I'll try again and update the code accordingly.
```
this.socket = net.createConnection(addr);
```

File: src/client.js (hunk 11 above)
Comment (asticode) [reply src/client.js:16]: OK makes sense. Once the last issues are resolved, I'll test this PR out and see if showing a dialog is actually the best option.

File: src/client.js (hunk 10 above)
Comment (shrey-gang) [reply src/client.js]: Pardon me for late response, I tried testing client code with below function it works very well with darwin and linux but somehow does not work in windows. The app gets installed but does not runs so debugging that is taking a little time. I did try the other approaches (i.e "tcp" check as you and the isIP() check) but somehow windows app is not working.

```
this.socket = net.createConnection(addr);
```

Below are the two approaches I have used to build the app, code snippet of bundler.json
```
"environments": [
  {"arch": "amd64", "os": "darwin"},
  {
    "arch": "amd64",
    "os": "windows",
    "env": {
      "CC": "x86_64-w64-mingw32-gcc",
      "CXX": "x86_64-w64-mingw32-g++",
      "CGO_ENABLED": "1"
    }
  }
]
```

and 

```
"environments": [
  {"arch": "amd64", "os": "linux"},
  {"arch": "386", "os": "windows"}
]
```

I need a bit more time to resolve this issue.

File: src/client.js (hunk 11 above)
Comment (shrey-gang) [reply src/client.js:16]: @asticode please test this PR

> OK makes sense. Once the last issues are resolved, I'll test this PR out and see if showing a dialog is actually the best option.

**Comment by shrey-gang** [comment]
@asticode 
Hi,
Apologies for not being active for quite some time.
Any update on this one, please.
I think all the work is done in this scope.
Please let me know if I missed something or there are some more changes required.

**Comment by asticode** [comment]
@shrey-gang could you resolve conflicts with changes that were made in between ? It seems `src/client.js` has conflicts

**Comment by shrey-gang** [comment]
> @shrey-gang could you resolve conflicts with changes that were made in between ? It seems `src/client.js` has conflicts

Conflict resolved

**Comment by gjchentw** [comment]
Hi @shrey-gang and @asticode ,

this PR changed behavior of init(addr) in client.js, that addr is no more a string and will broke start() in index.js takes process.argv[2] from go-astilectron binding. I tried to fix like this:

(https://github.com/gjchentw/astilectron/commit/f366088c0fb06ec3c2c2c5ade9821bde15813ae4)

```
        if(typeof addr === 'string') {
            let u = url.parse("tcp://" + addr, false, false);
            this.socket = new net.Socket();
            this.socket.connect(u.port, u.hostname, function() {});
        } else {
            // take addr as options of net.createConnection()
            this.socket = net.createConnection(addr);
        }
```
and make it worked with go-astilectron.

@shrey-gang please consider take this snippet or pull from (https://github.com/gjchentw/astilectron/commit/f366088c0fb06ec3c2c2c5ade9821bde15813ae4)

Thank you.
